### ISSUE TITLE
The callsite of f90_mmul_* flang runtime functions wrongly generated

### ISSUE THREAD

**Issue by pawosm-arm** [flang-compiler/flang#1005]
@mleair this bug is extremely hard to track and causes random errors and is very fragile to whatever was encountered on the stack. Also I still don't have a good minimal reproducer that would actually crash. Let me try to describe the problem anyway.

Consider the following program:
```
MODULE kinds
  INTEGER, PARAMETER :: dp = KIND(0.0D0)
END MODULE

program matmu
  use kinds
  integer, parameter :: m = 10, n = 20
  REAL(KIND = dp), DIMENSION(1:n, 1:n, 0:n)  :: hmat, umat

  hmat = 1._dp
  umat = -1._dp
  print *, hmat, umat
  call do_matmul(hmat, umat, 3, 3, 1)

contains

  subroutine do_matmul(hmat, umat, m, n, l)
    use kinds
    REAL(KIND = dp), DIMENSION(:, :, 0:), INTENT(IN)     :: hmat, umat
    integer, intent(in) :: m, n, l
    REAL(KIND = dp), ALLOCATABLE, DIMENSION(:, :)        :: a, tmp1, tmp2

    allocate(a(n, n))
    a(1:m, 1:m) = MATMUL(TRANSPOSE(umat(1:n, 1:m, l)), MATMUL(hmat(1:n, 1:n, l), umat(1:n, 1:m, l)))
    print *, a(1, 1)
    deallocate(a)
  end subroutine do_matmul

end program matmu
```

When compiled and executed, it prints what we expect and completes succesfully.
Yet when examined carefully in gdb, it turns out, something went wrong unnoticed under the hood:
```
(gdb) b f90_mmul_real8
Breakpoint 1 at 0x400af4
(gdb) r
...
Breakpoint 1, f90_mmul_real8 (ta=0, tb=0, mra=3, ncb=3, kab=3, alpha=0x401db0 <.C292_matmu_do_matmul>, a=0x4597c0, lda=3, b=0x459830, ldb=281474976690608, beta=0x401da8 <.C291_matmu_do_matmul>, c=0x45a4e0, ldc=3) at runtime/flang/mmreal8.c:58
58        int small_size = SMALL_ROWSA * SMALL_ROWSB * SMALL_COLSB;
(gdb)
```

Look at the `ldb` value and compare it to what `lda` and `ldc` have. It is definitely wrong, however in most cases it does not hurt. Until it hurts like that:
```
Thread 1 "cp2k.psmp" received signal SIGSEGV, Segmentation fault.
0x0000ffffbd50ab74 in f90_mmul_real8 (ta=0, tb=0, mra=3, ncb=3, kab=3,
    alpha=0x2ed5840 <.C292_atom_utils_atom_solve_>, a=0xc7d0c20, lda=3,
    b=0xc7d0c90, ldb=34411059240,
    beta=0x2ed5870 <.C291_atom_utils_atom_solve_>, c=0xc7d0d00, ldc=3)
    at runtime/flang/mmreal8.c:133
133                   temp += buffera[k] * b[bndx++];
```

I've tried to examine a call to `f90_mmul_real8` in the generated LLVM-IR and I've found that the passed value is never initialized!:
```
  %z_e_129_358 = alloca i64, align 8
.
.
.
  %452 = load i64, i64* %z_e_129_358, align 8
.
.
.
  %465 = bitcast void ()* @f90_mmul_real8 to void (i32, i32, i64, i64, i64, i8*, i8*, i64, i8*, i64, i8*, i8*, i64)*
  call void %465(i32 0, i32 0, i64 %426, i64 %429, i64 %432, i8* %433, i8* %435, i64 %438, i8* %451, i64 %452, i8* %453, i8* %461, i64 %464)
```

What do you think about that? Can you point me at the place where the variables for this call are prepared for being emitted to the IR?

**Comment by pawosm-arm** [comment]
The `z_e_129` symbol exists in `.ilm` and `.stb` files, so the problem must of been generated at `flang1` stage, therefore much earlier than I expected.

**Comment by pawosm-arm** [comment]
I've been analyzing `mmul_arg(int arr, int transpose, MMUL *mm)` function code in the `tools/flang1/flang1exe/func.c` file. What I figured out is that `ldim` value passed as `lda` and `ldc` is prepared as such:
```
#ifdef NOEXTENTG
  else if (HCCSYMG(sptr) && SCG(sptr) == SC_LOCAL && ALLOCG(sptr) &&
           (NOEXTENTG(sptr) || simply_contiguous(arr))) {
    /*
     * the EXTNTAST temp may not be defined for compiler-created
     * allocatable temps assigned the value of the argument.
     */
    ADSC *tad;
    tad = AD_DPTR(DTYPEG(sptr));
    ldim = mk_extent_expr(AD_LWBD(tad, 0), AD_UPBD(tad, 0));
  }
#endif
```
While `ldim` passed as `ldb` is prepared as such:
```
  if (STYPEG(sptr) == ST_MEMBER) {
    ....... }
  else ///////// HERE:
    ldim = ADD_EXTNTAST(DTYPEG(sptr), 0);
```
I've replaced `ldim = ADD_EXTNTAST(DTYPEG(sptr), 0);` with:
```
    ADSC *tad;
    tad = AD_DPTR(DTYPEG(sptr));
    ldim = mk_extent_expr(AD_LWBD(tad, 0), AD_UPBD(tad, 0));
```
...and now I'm observing the missing store operation after allocation and before read operation on the `z_e_...` variable passed as `ldb` parameter in the newly generated LLVM-IR. The value displayed by `gdb` was `20` this time, still wrong (kinda expected, as the replacement code isn't what it was supposed to be), but at least it's not a random value encountered on stack.
Still some hints needed.

**Comment by pawosm-arm** [comment]
I've been playing around with various ways of passing arguments to MATMUL where at least one of the arguments is the result of the other (nested) MATMUL. I've discovered that it's not only `ldb` which is wrongly passed! The two other strides (`lda`, `ldc`) can also be wrong, depending on which of the arguments (or both of them) are results of nested MATMULs. I've also noticed that whatever is passed as `lda` is also passed as `ldc` and when `lda` is wrong, `ldc` is equally wrong (and different than results of the same program compiled with gfortran)! The effect is usually a program that completes execution and producing wrong results (!!!!), or, if we're lucky, a segfault, which is the only way to be alarmed that something goes terribly wrong under the hood.

@mleair this is a serious bug which makes use of flang questionable for any serious applications!

**Comment by pawosm-arm** [comment]
@mleair we'll be working on fixing this, but your hints would be very helpful.

**Comment by mleair** [comment]
Looks like it is a problem with the 'a' argument. If you pass b,c,n to do_matmul() as dummy arguments but use the 'a' from the host (program mmul). then the code compiles and runs without issue. 

I suspect the bug is in the innermost call to matmul. For example, the following works for me:

     c(1:n, 1:n) =  matmul(a(1:n, 1:n), b(1:n, 1:n))

     c(1:n, 1:n) = matmul(matmul(matmul(a(1:n, 1:n), b(1:n, 1:n)), c), b(1:n, 1:n))


Not sure what is happening here without really digging into the generated code. You mentioned the temporary symbol z_e_129 above. Temporary symbols like z_e_129 are typically used in array bounds calculations. Maybe there is a conflict between the two uses of 'a' in the call-site; maybe there is an issue with bounds variables.

**Comment by pawosm-arm** [comment]
@mleair, I need to make you aware that MATMUL is implemented by calling one of the C functions from one of the two function families defined in the flang runtime: `f90_matmul_*` or `f90_mmul_*`. The choice from which function family the function needs to be taken seems to depend on the MATMUL parameters: for trivial cases `f90_matmul_*` is picked, otherwise `f90_mmul_*` is used.

The calls to `f90_matmul_*` functions seem to be generated correctly. By using just `nm` command, you can see what functions are called. Even in the example I've presented, the (easiest) innermost MATMULs are implemented by calling `f90_matmul_real8_i8`. But the outermost MATMUL is done by `f90_mmul_real8`.

In those two examples you've presented, all MATMULs are handed by calls to `f90_matmul_real8_i8`. Hence they all work properly.

The more curious example is with using host `a`, instead of passing it through the `do_matmul` subroutine argument. The outer matmul is still implemented by calling `f90_mmul_real8`, yet `gdb` shows that all the `lda`, `ldb`, `ldc` params have their correct values! This is strange... Need to investigate it more.

**Comment by pawosm-arm** [comment]
In the working case (where `a` is a host variable), compile-time constants are passed as `lda`, `ldb` and `ldc`. And it could have been expected in case of `a`: the host has all of the knowledge about `a` and its stride. Although the fact that the same (immediate) value is used for `ldb` and `ldc` is slightly suspicious, the focus should remain on the cases where all of the array variables are subroutine dummy parameters of the size (and stride) not known inside of the subroutine.

**Comment by pawosm-arm** [comment]
The more I look at what `func.c` code does, the more I see it is the other way round (comparing to what I've said two comments ago): it's the `f90_matmul_*` functions that can handle more complicated cases: they accept flang's array descriptors and can examine them. So I forced those functions to be picked for codegen whenever `NOEXTENTG` branch isn't taken in the `mmul_arg()` function quoted some comments ago. Surprisingly, with good results! At least with the code examples I have (cp2k originated). I'll try to follow this path further.

**Comment by pawosm-arm** [comment]
@mleair to summarize my latest findings: I've came up with two different modifications to `func.c` that for the cp2k-originated examples give correct outcomes. Now I need some help with proving which one of them (if any!) is correct (or maybe both are?).
First I've described in the comment above: whenever `else` branch is taken, give up on `f90_mmul_*` call preparations which in effect results in falling back into `f90_matmul_*` call:
```
diff --git a/tools/flang1/flang1exe/func.c b/tools/flang1/flang1exe/func.c
index 530b7b50..6c88c38e 100644
--- a/tools/flang1/flang1exe/func.c
+++ b/tools/flang1/flang1exe/func.c
@@ -7171,59 +7171,58 @@ mmul_arg(int arr, int transpose, MMUL *mm)
     m = mk_binop(OP_SUB, ub, lb, astb.bnd.dtype);
     m = mk_binop(OP_ADD, m, stride, astb.bnd.dtype);
     mm->extent[i] = m;
   }
   /* ldim must be before any tranpose */
   if (STYPEG(sptr) == ST_MEMBER) {
     ldim = ADD_EXTNTAST(DTYPEG(sptr), 0);
     ldim = check_member(mm->addr, ldim);
   }
 #ifdef NOEXTENTG
   else if (HCCSYMG(sptr) && SCG(sptr) == SC_LOCAL && ALLOCG(sptr) &&
            (NOEXTENTG(sptr) || simply_contiguous(arr))) {
     /*
      * the EXTNTAST temp may not be defined for compiler-created
      * allocatable temps assigned the value of the argument.
      */
     ADSC *tad;
     tad = AD_DPTR(DTYPEG(sptr));
     ldim = mk_extent_expr(AD_LWBD(tad, 0), AD_UPBD(tad, 0));
   }
 #endif
 #ifdef CONTIGATTRG
   else if (CONTIGATTRG(sptr)) {
     ADSC *tad;
     tad = AD_DPTR(DTYPEG(sptr));
     ldim = mk_extent_expr(AD_LWBD(tad, 0), AD_UPBD(tad, 0));
   }
 #endif
-  else {
-    ldim = ADD_EXTNTAST(DTYPEG(sptr), 0);
-  }
+  else
+    return FALSE;
   if (transpose) {
     /*  extents are post-tranposed */
     m = mm->extent[0];
     mm->extent[0] = mm->extent[1];
     mm->extent[1] = m;
   }
   if (astb.bnd.dtype != DT_INT8) {
     ldim = mk_convert(ldim, DT_INT8);
     for (i = 0; i < mm->rank; i++) {
       mm->extent[i] = mk_convert(mm->extent[i], DT_INT8);
     }
   }
   if (mm->rank == 1)
     mm->extent[1] = astb.k1;
   mm->ldim = ldim;
   return TRUE;
 }
```
This solution feels like something we're looking for, yet what worries me is that for some reason the original author of this code didn't do something that obvious: there must have been a reason for this.

The second idea is to stick to the calling of `f90_mmul_*` functions, but also to make sure that the AD (array descriptor) is examined and the stride is actually computed:
```
diff --git a/tools/flang1/flang1exe/func.c b/tools/flang1/flang1exe/func.c
index 530b7b50..14c0ce5e 100644
--- a/tools/flang1/flang1exe/func.c
+++ b/tools/flang1/flang1exe/func.c
@@ -7180,41 +7180,43 @@ mmul_arg(int arr, int transpose, MMUL *mm)
 #ifdef NOEXTENTG
   else if (HCCSYMG(sptr) && SCG(sptr) == SC_LOCAL && ALLOCG(sptr) &&
            (NOEXTENTG(sptr) || simply_contiguous(arr))) {
     /*
      * the EXTNTAST temp may not be defined for compiler-created
      * allocatable temps assigned the value of the argument.
      */
     ADSC *tad;
     tad = AD_DPTR(DTYPEG(sptr));
     ldim = mk_extent_expr(AD_LWBD(tad, 0), AD_UPBD(tad, 0));
   }
 #endif
 #ifdef CONTIGATTRG
   else if (CONTIGATTRG(sptr)) {
     ADSC *tad;
     tad = AD_DPTR(DTYPEG(sptr));
     ldim = mk_extent_expr(AD_LWBD(tad, 0), AD_UPBD(tad, 0));
   }
 #endif
   else {
-    ldim = ADD_EXTNTAST(DTYPEG(sptr), 0);
+    ADSC *tad;
+    tad = AD_DPTR(DTYPEG(sptr));
+    ldim = mk_extent_expr(AD_LWBD(tad, 0), AD_UPBD(tad, 0));
   }
   if (transpose) {
     /*  extents are post-tranposed */
     m = mm->extent[0];
     mm->extent[0] = mm->extent[1];
     mm->extent[1] = m;
   }
   if (astb.bnd.dtype != DT_INT8) {
     ldim = mk_convert(ldim, DT_INT8);
     for (i = 0; i < mm->rank; i++) {
       mm->extent[i] = mk_convert(mm->extent[i], DT_INT8);
     }
   }
   if (mm->rank == 1)
     mm->extent[1] = astb.k1;
   mm->ldim = ldim;
   return TRUE;
 }
```
The obvious problem with that second approach is that exactly the same thing is happening in two other `if` branches, where there is certain that the data are contiguous (either by `simply_contiguous(arr)` returning TRUE, which, as I checked, is not the case in our examples, or by `(CONTIGATTRG(sptr))` being TRUE, which isn't the case either). The second approach then doesn't feel right, even if in effect it adds those missing stack variable population code resulting in strides being passed to `f90_mmul_*` calls.

**Comment by pawosm-arm** [comment]
```

  retval = mmul(func_ast, func_args, lhs);
  if (retval >= 0)
    return retval;
```

**Comment by pawosm-arm** [comment]
Oh, I didn't notice the pinnacle of it all. In `mmul()` function there is:
```
  if (XBIT(47, 0x10000000))
    return -1;
```
The `xflag.n` file describes it as such:
```
0x10000000 Do not call our streamlined/dgemm-like matmul run-time routines.
```
Feels like I've been struggling with the obvious.

**Comment by mleair** [comment]
Can you just check if the arguments are assumed shape to decide which matmul routine to call?

**Comment by pawosm-arm** [comment]
@mleair I assume you had a change like the quoted below in mind:

```
diff --git a/tools/flang1/flang1exe/func.c b/tools/flang1/flang1exe/func.c
index 9c661ea3..8229b3de 100644
--- a/tools/flang1/flang1exe/func.c
+++ b/tools/flang1/flang1exe/func.c
@@ -7115,41 +7115,32 @@ mmul_arg(int arr, int transpose, MMUL *mm)
   int rank, dt, i;
   int lb, ub, stride;
   int m;

   sptr = find_array(arr, NULL);
   if (POINTERG(sptr)
 #ifdef CONTIGATTRG
       && !CONTIGATTRG(sptr)
 #endif
   )
     return FALSE;
   shape = A_SHAPEG(arr);
   if (!shape)
     return FALSE;
   mm->rank = SHD_NDIM(shape);
-  if (ASSUMSHPG(sptr) && mm->rank != 1
-#ifdef CONTIGATTRG
-      && !CONTIGATTRG(sptr)
-#endif
-  ) {
-    /*
-     * assumed-shaped arrays are guaranteed to be stride 1 in
-     * just the first dimension.
-     */
+  if (ASSUMSHPG(sptr))
     return FALSE;
-  }
   if (A_TYPEG(arr) == A_ID) {
     /*  whole */
     mm->addr = arr;
   } else if (A_TYPEG(arr) == A_MEM) {
     /*  whole -- allowing unsubscripted members is new as of 5/25/2012;
      *  so to back out, just add 'return FALSE;' here.
      */
     mm->addr = arr;
   } else if (mmul_array(arr)) {
     int asd;
     int subscr[MAXSUBS];
     asd = A_ASDG(arr);
     rank = ASD_NDIM(asd);
     for (i = 0; i < rank; ++i) {
       int ss;
```
Basically, give up on `f90_mmul_*` call preparations whenever `ASSUMSHPG(sptr)` is true.

Sadly, this didn't help.

**Comment by pawosm-arm** [comment]
Addressed by #1011
